src/coprocessor/codec/mysql/time.rs

```diff
@@ -30,6 +30,192 @@ const ZERO_DATE_STR: &'static str = "0000-00-00";
+    ///         otherwise Sunday.
+    /// when zero_week == true, week is from 0 to 53, otherwise from 1 to 53.
+    /// when first_weekday == true, the week that contains the first 'first-day-of-week' is week 1,
+    ///         otherwise weeks are numbered according to ISO 8601:1988.
+    fn calc_year_week(
+        &self,
+        monday_first: bool,
+        zero_week: bool,
+        first_weekday: bool,
+    ) -> (i32, i32) {
+        let week: i32;
+        let mut year = self.year();
+
+        let daynr = calc_day_number(self.year(), self.month() as i32, self.day() as i32);
+        let mut first_daynr = calc_day_number(self.year(), 1, 1);
+        let mut weekday = calc_weekday(first_daynr, !monday_first);
+        let mut week_year = !zero_week;
+        let mut days: i32;
+
+        if self.month() == 1 && (self.day() as i32) <= 7 - weekday {
+            if !week_year && ((first_weekday && weekday != 0) || (!first_weekday && weekday >= 4)) {
+                return (year, 0);
+            }
+            week_year = true;
+            year -= 1;
+            days = calc_days_in_year(year);
+            first_daynr -= days;
+            weekday = (weekday + 53 * 7 - days) % 7;
+        }
+
+        if (first_weekday && weekday != 0) || (!first_weekday && weekday >= 4) {
+            days = daynr - (first_daynr + 7 - weekday);
+        } else {
+            days = daynr - (first_daynr - weekday);
+        }
+
+        if week_year && days >= 52 * 7 {
+            weekday = (weekday + calc_days_in_year(year)) % 7;
+            if (!first_weekday && weekday < 4) || (first_weekday && weekday == 0) {
+                year += 1;
+                return (year, 1);
+            }
+        }
+        week = days / 7 + 1;
+        (year, week)
+    }
+
+    fn week(&self, monday_first: bool, zero_week: bool, first_weekday: bool) -> i32 {
+        if self.month() == 0 || self.day() == 0 {
+            return 0;
+        }
+        let (_, week) = self.calc_year_week(
+            monday_first,
+            zero_week,
+            if monday_first { false } else { first_weekday },
+        );
+        week
+    }
+}
+
+fn abbr_day_of_month(day: i32) -> &'static str {
+    match day {
+        1 | 21 | 31 => "st",
+        2 | 22 => "nd",
+        3 | 23 => "rd",
+        _ => "th",
+    }
+}
+
+// calculates days since 0000-00-00.
+fn calc_day_number(year: i32, month: i32, day: i32) -> i32 {
+    if year == 0 && month == 0 {
+        return 0;
+    }
+    let mut year2 = year;
+    let mut delsum = 365 * year + 31 * (month - 1) + day;
+    if month <= 2 {
+        year2 -= 1;
+    } else {
+        delsum -= (month * 4 + 23) / 10;
+    }
+    let temp = ((year2 / 100 + 1) * 3) / 4;
+    delsum + year2 / 4 - temp
+}
+
+/// calculates days in one year, it works with 0 <= year <= 99.
+fn calc_days_in_year(year: i32) -> i32 {
+    if (year & 3) == 0 && (year % 100 != 0 || (year % 400 == 0 && (year != 0))) {
```

Comment:
It seems wouldn't benefit performance and ugly to understand? @tiancaiamao

Reply:
Yeah, it works but nobody know why.